### PR TITLE
feat (hatchery/swarm): opts --ports/privileged/volume

### PR DESCRIPTION
**Pull request by yesnault**
allowed only on hatchery not 'shared.infra'

options have to be setted on Job Prerequisite: options on worker model

Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

### CODE REVIEW

**Comment by yesnault** [comment]
(wip, do not merge yet)

File: engine/hatchery/swarm/swarm.go [hunk 1]
```diff
@@ -404,6 +441,65 @@ func (h *HatcherySwarm) createAndStartContainer(name, image, network, networkAli
...
+					continue // it's image name
+				}
+				if isSharedInfra {
+					return nil, fmt.Errorf("You could not use this docker options '%s' with a 'shared.infra' hatchery. Please use you own hatchery or remove this option.", opt)
+				}
+				if strings.HasPrefix(opt, "--port=") {
+					if err := dockerOpts.computeDockerOptsPorts(opt); err != nil {
+						return nil, err
+					}
+				} else if opt == "--priviledge" {
```
Comment (fsamin) [inline engine/hatchery/swarm/swarm.go]: should be `--privileged `

File: engine/hatchery/swarm/swarm.go [hunk 2]
```diff
@@ -404,6 +441,65 @@ func (h *HatcherySwarm) createAndStartContainer(name, image, network, networkAli
...
+}
+
+func (d *dockerOpts) computeDockerOptsPorts(arg string) error {
+	if regexPort.MatchString(arg) {
+		s := regexPort.FindStringSubmatch(arg)
+		//s = --port=8081:8182/tcp // hostPort:containerPort
+		//s[0] = --port=8081:8182/tcp
+		//s[1] = 8081 // hostPort
+		//s[2] = 8182/tcp  // containerPort
+		containerPort := s[2]
```
Comment (fsamin) [inline engine/hatchery/swarm/swarm.go]: is could be nice to be tcp per default (id s[2] has not suffix /.*, add /tcp)

File: engine/hatchery/swarm/swarm.go [hunk 3]
```diff
@@ -362,30 +386,43 @@ func (h *HatcherySwarm) createNetwork(name string) error {
...
+	name, image, network, networkAlias string
+	cmd, env                           []string
+	labels                             map[string]string
+	memory                             int64
+	dockerOpts                         *dockerOpts
+}
+
 //shortcut to create+start(=run) a container
-func (h *HatcherySwarm) createAndStartContainer(name, image, network, networkAlias string, cmd, env []string, labels map[string]string, memory int64) error {
+func (h *HatcherySwarm) createAndStartContainer(cArgs containerArgs) error {
```
Comment (fsamin) [inline engine/hatchery/swarm/swarm.go]: add this comment: `You could not use this docker options '%s' with a 'shared.infra' hatchery`

File: engine/hatchery/swarm/swarm.go [hunk 4]
```diff
@@ -273,7 +274,15 @@ func (h *HatcherySwarm) SpawnWorker(spawnArgs hatchery.SpawnArguments) (string,
 					"hatchery":       h.Config.Name,
 				}
 				//Start the services
-				if err := h.createAndStartContainer(serviceName, img, network, r.Name, []string{}, env, labels, serviceMemory); err != nil {
+				if err := h.createAndStartContainer(containerArgs{
```
Comment (bnjjj) [inline engine/hatchery/swarm/swarm.go]: not very readable

File: engine/hatchery/swarm/swarm.go [hunk 5]
```diff
@@ -335,8 +344,23 @@ func (h *HatcherySwarm) SpawnWorker(spawnArgs hatchery.SpawnArguments) (string,
...
 	}
 
+	dockerOpts, errDockerOpts := computeDockerOpts(h.hatch.IsSharedInfra, spawnArgs.Requirements)
+	if errDockerOpts != nil {
+		return name, errDockerOpts
+	}
+
 	//start the worker
-	if err := h.createAndStartContainer(name, spawnArgs.Model.Image, network, "worker", cmd, env, labels, memory); err != nil {
+	if err := h.createAndStartContainer(containerArgs{
```
Comment (bnjjj) [inline engine/hatchery/swarm/swarm.go]: same

File: engine/hatchery/swarm/swarm.go (hunk 2 above)
Comment (yesnault) [reply engine/hatchery/swarm/swarm.go]: done

File: engine/hatchery/swarm/swarm.go (hunk 3 above)
Comment (yesnault) [reply engine/hatchery/swarm/swarm.go]: done

**Comment by yesnault** [comment]
tested, ready to merge;

File: engine/hatchery/swarm/swarm.go (hunk 2 above)
Comment (yesnault) [reply engine/hatchery/swarm/swarm.go]: here: https://github.com/ovh/cds/pull/1481/commits/e8a30ef6304099bac78a8d7bdfafd25e10945b57#diff-5b928cb03058ec28cb0709fcac0a56b6R497

File: engine/hatchery/swarm/swarm.go (hunk 3 above)
Comment (yesnault) [reply engine/hatchery/swarm/swarm.go]: here https://github.com/ovh/cds/pull/1481/commits/e8a30ef6304099bac78a8d7bdfafd25e10945b57#diff-5b928cb03058ec28cb0709fcac0a56b6R219

**Comment by yesnault** [comment]
capture d ecran 2017-11-18 a 18 33 32

**Comment by yesnault** [comment]
do not merge, I'll add volume on this PR.

**Comment by yesnault** [comment]
Tested with volume, ready to merge

**Pull request merged**